File: DataStructures/Queue/Python/queue.py
from typing import TypeVar, Generic

T = TypeVar('T')

class Queue(Generic[T]):

    def __init__(self) -> None:
        self.items: list[T] = []
    
    def enqueue(self, item: T) -> None:
        self.items.append(item)

    def dequeue(self) -> T:
        return self.items.pop(0)
    
    def is_empty(self) -> bool:
        return len(self.items) == 0
    
    def peek(self) -> T | None:
        if not self.is_empty:
            return self.items[0]
        return None
    
    def size(self) -> int:
        return len(self.items)
    
    def to_string(self) -> str:
        return str(self.items)

File: DataStructures/Queue/Javascript/queue.mjs
// run command: npm run js-queue

export class Queue {
    items;

    constructor() {
        this.items = [];
    }

    isEmpty() {
        return this.items.length === 0;
    }

    enqueue(item) {
        this.items = [...this.items, item];
    }

    dequeue() {
        if (this.isEmpty()) {
            return null;
        }
        const item = this.items[0];
        this.items = this.items.slice(1);
        return item;
    }

    peek() {
        if (this.isEmpty()) {
            return null;
        }
        return this.items[0];
    }

    size() {
        return this.items.length;
    }
}

const queue = new Queue();
console.log(queue.isEmpty());
queue.enqueue('a');
queue.enqueue('b');
console.log(queue.isEmpty());
console.log(queue.size());
console.log(queue.peek());
console.log(queue.dequeue());
console.log(queue.peek());
